### ISSUE TITLE
Dynamic loading broken with Perl 5.36

### ISSUE THREAD

**Issue by foutrelis** [vim/vim#10512]
### Steps to reproduce

1. Build vim with `--enable-perlinterp=dynamic` against Perl 5.36
2. Inspect the vim binary with `readelf -d` revealing a DT_NEEDED entry for libperl.so

### Expected behaviour

The vim binary should not be linked to libperl so it can be dynamically loaded at runtime.

Comparing the `readelf -Ws vim | grep UND` output between Perl 5.34 and 5.36 shows a new symbol reference: `PL_current_context`

### Version of Vim

8.2.5046

### Environment

Arch Linux x86_64

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> 1. Build vim with `--enable-perlinterp=dynamic` against Perl 5.36
> 2. Inspect the vim binary with `readelf -d` revealing a DT_NEEDED entry for libperl.so
> 
> ### Expected behaviour
> 
> The vim binary should not be linked to libperl so it can be dynamically loaded at runtime.
> 
> Comparing the `readelf -Ws vim | grep UND` output between Perl 5.34 and 5.36 shows a new symbol reference: `PL_current_context`
> 
> ### Version of Vim
> 
> 8.2.5046
> 
> ### Environment
> 
> Arch Linux x86_64

My Ubuntu system has Perl 5.34 and it looks OK.  Is this a problem
specific for Perl 5.36 or for Arch Linux?

-- 
WOMAN:   Well, 'ow did you become king then?
ARTHUR:  The Lady of the Lake, [angels sing] her arm clad in the purest
         shimmering samite, held aloft Excalibur from the bosom of the water
         signifying by Divine Providence that I, Arthur, was to carry
         Excalibur.  [singing stops] That is why I am your king!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by foutrelis** [comment]
> Is this a problem specific for Perl 5.36 or for Arch Linux?

It appears to be specific to Perl 5.36 because it was fine with Perl 5.34 on Arch Linux too.

As far as I can tell, the `PL_current_context` symbol is new in Perl 5.36 and was introduced by https://github.com/Perl/perl5/commit/3683a9ee5ad61f8cb5eac30e61b4936bd0445bdb.

**Comment by brammool** [comment]
Vim does not use PL_current_context, I'm not sure if we can do anything about this.
You could ask the Perl maintainers about this.

**Comment by foutrelis** [comment]
As far as I can see, including `perl.h` brings in Perl's `thread.h` which in turn contains a reference to `PL_current_context`.

A code comment added as part of https://github.com/Perl/perl5/commit/3683a9ee5ad61f8cb5eac30e61b4936bd0445bdb has this to say about it:
```
 * This is not a regular "global" in that we don't know whether it needs to
 * exist until we include threads.h, and we don't want it as part of any
 * global struct (if that or something similar is re-introduced.
```

This feels similar to how previous major Perl upgrades have required changes in Vim to accommodate them, even though Vim's code didn't change in regard to its Perl usage.

**Comment by brammool** [comment]
> As far as I can see, including `perl.h` brings in Perl's `thread.h` which in turn contains a reference to `PL_current_context`.
> 
> A code comment added as part of https://github.com/Perl/perl5/commit/3683a9ee5ad61f8cb5eac30e61b4936bd0445bdb has this to say about it:
> ```
>  * This is not a regular "global" in that we don't know whether it needs to
>  * exist until we include threads.h, and we don't want it as part of any
>  * global struct (if that or something similar is re-introduced.
> ```
> 
> This feels similar to how previous major Perl upgrades have required
> changes in Vim to accommodate them, even though Vim's code didn't
> change in regard to its Perl usage.

Perhaps this is similar to how PL_thr_key is handled?

-- 
Q: How does a UNIX Guru pick up a girl?
A: look; grep; which; eval; nice; uname; talk; date;

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
> This following change prevents Vim from linking to libperl (without breaking the test suite):
> 
> ```diff
> diff --git a/src/if_perl.xs b/src/if_perl.xs
> index 1afb1d4cf..8aab5337b 100644
> --- a/src/if_perl.xs
> +++ b/src/if_perl.xs
> @@ -168,6 +168,10 @@ typedef int XSUBADDR_t;
>  typedef int perl_key;
>  # endif
>  
> +#if defined(USE_ITHREADS) && defined(PERL_THREAD_LOCAL)
> +PERL_THREAD_LOCAL void *PL_current_context;
> +#endif
> +
>  # ifndef MSWIN
>  #  include <dlfcn.h>
>  #  define HANDLE void*
> ```
> 
> However, I am not sure if this is the correct approach or how it interacts with the same symbol in libperl:
> 
> ```
> $ readelf -Ws src/vim | grep PL_current_context
>   1184: 0000000000000000     8 TLS     GLOBAL DEFAULT   17 PL_current_context
> 
> $ readelf -Ws /usr/lib/perl5/5.36/core_perl/CORE/libperl.so | grep PL_current_context
>   1452: 0000000000000000     8 TLS     GLOBAL DEFAULT   15 PL_current_context
> ```

This will create a duplicate variable, that can't work properly.

-- 
Courtroom Quote #19:
Q:  Doctor, how many autopsies have you performed on dead people?
A:  All my autopsies have been performed on dead people.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
> just add the extra symbol to be dlsym-ed (just like PL_thr_key) can call it a day

can you create a PR for this please 🙏 ?

**Comment by k-takata** [comment]
This should have been fixed in 9.0.1700. Closing.
